bugsnag-android-core/src/test/java/com/bugsnag/android/ReportDeliveryDelegateTest.kt

```diff
@@ -0,0 +1,87 @@
+package com.bugsnag.android
+
+import com.bugsnag.android.BugsnagTestUtils.generateImmutableConfig
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotNull
+import org.junit.Assert.fail
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mock
+import org.mockito.junit.MockitoJUnitRunner
+import java.util.Date
+
+@RunWith(MockitoJUnitRunner::class)
+internal class ReportDeliveryDelegateTest {
+
+    @Mock
+    lateinit var eventStore: EventStore
+
+    val config = generateImmutableConfig()
+    val breadcrumbState = BreadcrumbState(50, NoopLogger)
+    private val logger = InterceptingLogger()
+    lateinit var deliveryDelegate: ReportDeliveryDelegate
+    val handledState = HandledState.newInstance(HandledState.REASON_UNHANDLED_EXCEPTION)
+    val event = Event(RuntimeException("Whoops!"), config, handledState)
+
+    @Before
+    fun setUp() {
+        deliveryDelegate = ReportDeliveryDelegate(logger, eventStore, config, breadcrumbState)
+        event.session = Session("123", Date(), User(null, null, null), false)
+    }
+
+    @Test
+    fun generateUnhandledReport() {
+        var msg: StateEvent.NotifyUnhandled? = null
+        deliveryDelegate.addObserver { _, arg ->
+            msg = arg as StateEvent.NotifyUnhandled
+        }
+        deliveryDelegate.deliverEvent(event)
+
+        // verify message sent
+        assertNotNull(msg)
+
+        // check session count incremented
+        assertEquals(1, event.session!!.unhandledCount)
+        assertEquals(0, event.session!!.handledCount)
+    }
+
+    @Test
+    fun generateHandledReport() {
+        val state = HandledState.newInstance(HandledState.REASON_HANDLED_EXCEPTION)
+        val event = Event(RuntimeException("Whoops!"), config, state)
+        event.session = Session("123", Date(), User(null, null, null), false)
+
+        var msg: StateEvent.NotifyHandled? = null
+        deliveryDelegate.addObserver { _, arg ->
+            msg = arg as StateEvent.NotifyHandled
+        }
+        deliveryDelegate.deliverEvent(event)
+
+        // verify message sent
+        assertNotNull(msg)
+
+        // check session count incremented
+        assertEquals(0, event.session!!.unhandledCount)
+        assertEquals(1, event.session!!.handledCount)
+    }
+
+    @Test
+    fun deliverReport() {
```

Comment:
Consider tests for events with no/empty errors - impact on breadcrumbs

Reply:
I've added a test to ensure that if an Event has no errors the payload is dropped before attempting delivery.